src/Microsoft.AspNetCore.Razor.Tools/GenerateCommand.cs

```diff
@@ -0,0 +1,204 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.VisualStudio.LanguageServices.Razor;
+using Newtonsoft.Json;
+
+namespace Microsoft.AspNetCore.Razor.Tools
+{
+    internal class GenerateCommand : CommandBase
+    {
+        public GenerateCommand(Application parent)
+            : base(parent, "generate")
+        {
+            Sources = Argument("sources", ".cshtml files to compile", multipleValues: true);
+            ProjectDirectory = Option("-p", "project root directory", CommandOptionType.SingleValue);
+            OutputDirectory = Option("-o", "output directory", CommandOptionType.SingleValue);
+            TagHelperManifest = Option("-t", "tag helper manifest file", CommandOptionType.SingleValue);
+        }
+
+        public CommandArgument Sources { get; }
+
+        public CommandOption OutputDirectory { get; }
+
+        public CommandOption ProjectDirectory { get; }
+
+        public CommandOption TagHelperManifest { get; }
+
+        protected override Task<int> ExecuteCoreAsync()
+        {
+            var result = ExecuteCore(
+                projectDirectory: ProjectDirectory.Value() ?? Environment.CurrentDirectory,
+                outputDirectory: OutputDirectory.Value(),
+                tagHelperManifest: TagHelperManifest.Value(),
+                sources: Sources.Values.ToArray());
+
+            return Task.FromResult(result);
+        }
+
+        protected override bool ValidateArguments()
+        {
+            if (string.IsNullOrEmpty(OutputDirectory.Value()))
+            {
+                Error.WriteLine($"{OutputDirectory.ValueName} not specified.");
```

Comment:
BTW when we run on the server, we need to replace these `TextWriters` with one hooked up to the server logger.

Reply:
Will address in https://github.com/aspnet/Razor/issues/1954.